content/posts/2021/11/heimdallr.md

```diff
@@ -0,0 +1,213 @@
+```c {linenos=true,hl_lines=[2,4,7]}
+if(rank == 0) {
+    MPI_Isend(buf, BUF_SIZE, MPI_DOUBLE, 1, 0, MPI_COMM_WORLD, &req);
+    for(int i = 0; i < BUF_SIZE; ++i)
+        buf[i] = 42.0;
+}
+else if(rank == 1) {
+    MPI_Recv(buf, BUF_SIZE, MPI_DOUBLE, 0, 0, MPI_COMM_WORLD, &status);
+}
+```
+
+In this example process 0 tries to send a buffer to process 1 using MPI's non-blocking send function `MPI_Isend`. 
+The non-blocking send operation in line 2 allows process 0 to continue its execution before the sending of the message has concluded.
+The problem arises in lines 3-4 where process 0 also modifies the contents of the data buffer that is being sent.
+Since the message passing process might still be running this may also modify the contents of the sent message and thereby cause a program
+error because this behavior was not intended by the programmer.
+
+This is a known safety issue with the use of non-blocking communication in MPI. A data buffer that is used in a non-blocking operation
+is in an 'unsafe' state until it has been made sure that the message passing operation on it has concluded. To check the status of a
+non-blocking operation an thereby the safety status of its data buffer MPI provides functions like `MPI_Wait` that blocks the 
+current process until the referenced message passing operation is confirmed to be finished. The MPI standard requires such a function
+to be called before accessing a data buffer again that has been used in non-blocking communication. Adding a `MPI_Wait` call
+between lines 2-3 of the example code would make this program work correctly.
+
+The problem with all of this is that MPI requires the programmer to always remember this behavior and neither the library nor the compiler
+are able to detect and warn users of potential errors with buffer safety for non-blocking communication.
+
+## Leveraging Rust's ownership for buffer safety
+
+The core concept of Rust's memory management is the so called 'ownership' feature. Ownership works in a way that every data object in Rust
+has exactly one owner. Once the owner variable goes out of scope the data is automatically deallocated. There can be references to an object
+but only within a limited rule-set. A variable can either have an unlimited number of immutable (read-only) references or exactly **one**
+mutable reference. These limitations allow the Rust compiler to reason about correct memory usage.
+
+```c {linenos=true,hl_lines=[2,3,8]}
+if client.id == 0 {
+    let handle = client.send_nb(buf, 1, 0)?;
+    buf = handle.data()?;
+    for i in 0..BUF_SIZE {
+        buf[i] = 42.0;
+    }
+} else if client.id == 1 {
+    buf = client.receive(0, 0)?;
+}
+```
+
+This is the heimdallr equivalent of the non-blocking MPI code that we have seen previously. The send operation in line 2 makes use of Rust's ownership
+concept to  protect the data buffer that is being sent. Since there can be only one owner of the `buf` variable passing it directly to a function call
```

Comment:
- Nitpick: Double space in "to  protect" (probably doesn't matter)
- Comma after "variable"